src/main/java/org/truffleruby/cext/ValueWrapperManager.java

```diff
@@ -191,22 +193,32 @@ public synchronized void addFreeBlock(long blockBase) {
 
     public static class HandleBlock {
 
-        public static final HandleBlock DUMMY_BLOCK = new HandleBlock(0, null);
+        public static final HandleBlock DUMMY_BLOCK = new HandleBlock(null, 0, null);
+
+        private static final Set<HandleBlock> keepAlive = new HashSet<>();
 
         private final long base;
         @SuppressWarnings("rawtypes") private final ValueWrapper[] wrappers;
         private int count;
 
-        public HandleBlock() {
-            this(allocator.getFreeBlock(), new ValueWrapper[BLOCK_SIZE]);
+        public HandleBlock(RubyContext context) {
+            this(context, allocator.getFreeBlock(), new ValueWrapper[BLOCK_SIZE]);
         }
 
-        private HandleBlock(long base, ValueWrapper[] wrappers) {
+        private HandleBlock(RubyContext context, long base, ValueWrapper[] wrappers) {
+            if (context != null && context.getOptions().CEXTS_KEEP_HANDLES_ALIVE) {
+                keepAlive(this);
+            }
             this.base = base;
             this.wrappers = wrappers;
             this.count = 0;
         }
 
+        @TruffleBoundary
+        private static void keepAlive(HandleBlock block) {
+            keepAlive.add(block);
```

Comment:
This looks like it needs synchronization.

Reply:
Fixed.